hazelcast/src/main/java/com/hazelcast/map/impl/query/ParallelAccumulationExecutor.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.query;
+
+import com.hazelcast.aggregation.Aggregator;
+import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.spi.serialization.SerializationService;
+import com.hazelcast.util.executor.ManagedExecutorService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.util.FutureUtil.RETHROW_EVERYTHING;
+import static com.hazelcast.util.FutureUtil.returnWithDeadline;
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+/**
+ * Implementation of the {@link AccumulationExecutor} that runs the accumulation in a multi-threaded way.
+ * Entries are split into chunks and each chunk is sent as a task to the underlying executor.
+ * At the end the result is merged to a single AggregationResult.
+ */
+public class ParallelAccumulationExecutor implements AccumulationExecutor {
+
+    private static final int THREAD_SPLIT_COUNT = 8;
+    private static final int ACCUMULATION_EXECUTION_TIMEOUT_MINUTES = 5;
+
+    private final ManagedExecutorService executor;
+    private final SerializationService serializationService;
+
+    public ParallelAccumulationExecutor(ManagedExecutorService executor, SerializationService serializationService) {
+        this.executor = executor;
+        this.serializationService = serializationService;
+    }
+
+    @SuppressWarnings("unchecked")
+    public AggregationResult execute(
+            Aggregator aggregator, Collection<QueryableEntry> entries, Collection<Integer> partitionIds) {
+        Collection<Aggregator> chunkAggregators = accumulateParallel(aggregator, entries);
+
+        Aggregator resultAggregator = clone(aggregator);
+        resultAggregator.onAccumulationFinished();
+
+        for (Aggregator chunkAggregator : chunkAggregators) {
+            resultAggregator.combine(chunkAggregator);
+        }
+        resultAggregator.onCombinationFinished();
```

Comment:
It doesn't really matter, since the user will get an exception and the resulgAggregator will be discarded.